src/segment_strava.go

```diff
@@ -0,0 +1,258 @@
+	props Properties
+	env   Environment
+
+	StravaData
+	ActivityIcon string
+	Ago          string
+	Hours        int
+}
+
+const (
+	StravaAccessToken  = "strava_access_token"
+	StravaRefreshToken = "strava_refresh_token"
+	StravaClientID     = "76033"
+	StravaClientSecret = "1111111111"
+	StravaApiURL       = "https://www.strava.com/api/v3/athlete/activities?page=1&per_page=1"
+)
+
+const (
+	// Your complete api URL and APIKey like this
+	InitialAccessToken  Property = "access_token"
+	InitialRefreshToken Property = "refresh_token"
+	RideIcon            Property = "ride_icon"
+	RunIcon             Property = "run_icon"
+	SkiingIcon          Property = "skiing_icon"
+	WorkOutIcon         Property = "workout_icon"
+	UnknownActivityIcon Property = "unknown_activity_icon"
+)
+
+// StravaData struct contains the API data
+type StravaData struct {
+	Type      string    `json:"type"`
+	StartDate time.Time `json:"start_date"`
+	Name      string    `json:"name"`
+	Distance  float64   `json:"distance"`
+	Duration  float64   `json:"moving_time"`
+}
+
+type TokenExchange struct {
+	TokenType    string `json:"token_type"`
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	ExpiresIn    int    `json:"expires_in"`
+	ExpiresAt    int    `json:"expires_at"`
+}
+
+func (ns *strava) enabled() bool {
+	data, err := ns.getResult()
+	if err != nil {
+		return false
+	}
+	ns.StravaData = *data
+	ns.ActivityIcon = ns.getActivityIcon()
+	ns.Hours = ns.getHours()
+	ns.Ago = ns.getAgo()
+
+	return true
+}
+
+func (ns *strava) getHours() int {
+	hours := time.Since(ns.StartDate).Hours()
+
+	return int(math.Floor(hours))
+}
+
+func (ns *strava) getAgo() string {
+	if ns.Hours > 24 {
+		days := int32(math.Floor(float64(ns.Hours) / float64(24)))
+		return fmt.Sprintf("%d", days) + string('d')
+	}
+	return fmt.Sprintf("%d", ns.Hours) + string("h")
+}
+
+func (ns *strava) getActivityIcon() string {
+	switch ns.Type {
+	case "Ride":
+		return ns.props.getString(RideIcon, "\uf5a2")
+	case "Run":
+		return ns.props.getString(RunIcon, "\ufc0c")
+	case "NordicSki":
+	case "AlpineSki":
+	case "BackcountrySki":
+		return ns.props.getString(SkiingIcon, "\ue213")
+	case "WorkOut":
+		return ns.props.getString(WorkOutIcon, "\ue213")
+	default:
+		return ns.props.getString(UnknownActivityIcon, "\ue213")
+	}
+	return ns.props.getString(UnknownActivityIcon, "\ue213")
+}
+
+func (ns *strava) string() string {
+	segmentTemplate := ns.props.getString(SegmentTemplate, "{{.Ago}}")
+	template := &textTemplate{
+		Template: segmentTemplate,
+		Context:  ns,
+		Env:      ns.env,
+	}
+	text, err := template.render()
+	if err != nil {
+		return err.Error()
+	}
+
+	return text
+}
+
+func (ns *strava) getAccessToken() (string, error) {
+	var at, rt string
+	var hasAt, hasRt bool
+	var err error
+
+	// get directly from cache
+	at, hasAt = ns.env.cache().get(StravaAccessToken)
+	if !hasAt {
+		rt, hasRt = ns.env.cache().get(StravaRefreshToken)
+		if hasRt {
```

Comment:
do we want to do this, or try the `access_token` and if that call fails, use the `refresh_token` and try again?